### PR TITLE
Remove KIR/PoC and introduce KCF/KFF from API returns and internal code 

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

This PR is a part of KGP-6 implementation, however, it causes a change in the API return format. 

1. The change in RewardSpec affects to `klay_getRewards` API
Before
```
{
        "minted": 6400000000000000000,
        "totalFee": 70825700000000000,
        "burntFee": 35412850000000000,
        "proposer": 3217706425000000000,
        "stakers": 0,
        "kgf": 2574165140000000000,
        "kir": 643541285000000000,
        "rewards": {
            "0x179679457f93094a4e7186abcb2089661e92fc22": 3217706425000000000,
            "0x278e6332d69eed782784d21802e3504a64a16456": 2574165140000000000,
            "0x3d803a7375a8ee5996f52a8d6725637a89f5bbf8": 643541285000000000
        }
```

After
```
{
        "minted": 6400000000000000000,
        "totalFee": 70825700000000000,
        "burntFee": 35412850000000000,
        "proposer": 3217706425000000000,
        "stakers": 0,
        "kff": 2574165140000000000,
        "kcf": 643541285000000000,
        "rewards": {
            "0x179679457f93094a4e7186abcb2089661e92fc22": 3217706425000000000,
            "0x278e6332d69eed782784d21802e3504a64a16456": 2574165140000000000,
            "0x3d803a7375a8ee5996f52a8d6725637a89f5bbf8": 643541285000000000
        }
```

2. The change in StakingInfo struct affects to `klay_getStakingInfo` API.
This change also affects to StakingInfo data stored in database, so the unmarshal function of StakingInfo is added to support both legacy data and new data. 

Before
```
{
        "BlockNum": 115171200,
        "CouncilNodeAddrs": [ ... ... ],
        "CouncilStakingAddrs": [ ... ... ],
        "CouncilRewardAddrs": [ ... ... ],
        "KIRAddr": "0x3d803a7375a8ee5996f52a8d6725637a89f5bbf8",
        "PoCAddr": "0x278e6332d69eed782784d21802e3504a64a16456",
        "UseGini": true,
        "Gini": 0.61,
        "CouncilStakingAmounts": [ ... ... ]
    }
```

After
```
{
        "blockNum": 115171200,
        "councilNodeAddrs":  [ ... ... ],
        "councilStakingAddrs":  [ ... ... ],
        "councilRewardAddrs":  [ ... ... ],
        "kcfAddr": "0x3d803a7375a8ee5996f52a8d6725637a89f5bbf8",
        "kffAddr": "0x278e6332d69eed782784d21802e3504a64a16456",
        "useGini": true,
        "gini": 0.61,
        "councilStakingAmounts":  [ ... ... ],
    }
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kjeom** [comment]
No need to add hard fork logic?

**Comment by blukat29** [comment]
`chaindata/misc` produced by this version may break old nodes. Do we need an intermediate period where both names are stored in the database?

File: reward/staking_info.go [hunk 1]
```diff
@@ -44,21 +44,66 @@ var (
...
+	type extendedSt struct {
+		BlockNum              uint64           `json:"blockNum"`
+		CouncilNodeAddrs      []common.Address `json:"councilNodeAddrs"`
+		CouncilStakingAddrs   []common.Address `json:"councilStakingAddrs"`
+		CouncilRewardAddrs    []common.Address `json:"councilRewardAddrs"`
+		KCFAddr               common.Address   `json:"kcfAddr"`
+		KFFAddr               common.Address   `json:"kffAddr"`
+		UseGini               bool             `json:"useGini"`
+		Gini                  float64          `json:"gini"`
+		CouncilStakingAmounts []uint64         `json:"councilStakingAmounts"`
```
Comment (blukat29) [inline reward/staking_info.go:75]: The implicit JSON tag was `"BlockNum"`. Therefore this code cannot read legacy staking info.

File: reward/staking_info_test.go [hunk 2]
```diff
@@ -390,3 +392,102 @@ func TestConsolidatedStakingInfo(t *testing.T) {
...
+		},
+		common.HexToAddress("0x673003e5f9a852d3dc85b83d16ef62d45497fb96"),
+		common.HexToAddress("0x576dc0c2afeb1661da3cf53a60e76dd4e32c7ab1"),
+		false,
+		-1,
+		[]uint64{5000000, 5000000, 5000000, 5000000},
+	}
+
+	// initialize StakingManager
+	_ = NewStakingManager(newTestBlockChain(), newDefaultTestGovernance(), nil)
```
Comment (blukat29) [inline reward/staking_info_test.go:468]:
```suggestion
	oldStakingManager := GetStakingManager()
	defer SetTestStakingManager(oldStakingManager)
```
NewStakingManager interferes with other tests.

File: reward/staking_info_test.go [hunk 3]
```diff
@@ -390,3 +392,102 @@ func TestConsolidatedStakingInfo(t *testing.T) {
...
+
+	for _, info := range []interface{}{oldInfo, newInfo} {
+		// reset database
+		stakingManager.stakingInfoDB = database.NewMemoryDBManager()
```
Comment (blukat29) [inline reward/staking_info_test.go:472]:
```suggestion
		SetTestStakingManagerWithDB(database.NewMemoryDBManager())
```
Do not modify the existing instance. Create a new one in each loop.

Comment (aidan-kwon) [reply reward/staking_info_test.go]: Applied. Thanks!

File: reward/staking_info_test.go (hunk 2 above)
Comment (aidan-kwon) [reply reward/staking_info_test.go]: Applied. Thanks!

**Review by toniya-klaytn — APPROVED** [review]

**Review by praveen-klaytn — DISMISSED** [review]

**Comment by aidan-kwon** [comment]
> No need to add hard fork logic?

This change affects to only two APIs that is not widely used. So, I just make a change and announce it by publishing an interface KIP.

**Comment by aidan-kwon** [comment]
> `chaindata/misc` produced by this version may break old nodes. Do we need an intermediate period where both names are stored in the database?

I introduced a JSON marshal method to provide backward compatibility

File: reward/staking_info.go [hunk 4]
```diff
@@ -44,21 +44,103 @@ var (
...
+	Gini    float64 `json:"gini"` // gini coefficient
 
 	// Derived from CouncilStakingAddrs
-	CouncilStakingAmounts []uint64 // Staking amounts of Council
+	CouncilStakingAmounts []uint64 `json:"councilStakingAmounts"` // Staking amounts of Council
+}
+
+// MarshalJSON supports json marshalling for both oldStakingInfo and StakingInfo
+// TODO-klaytn-Mantle: remove this marshal function when backward-compatibility for KIR/PoC is not needed
+func (st StakingInfo) MarshalJSON() ([]byte, error) {
```
Comment (ian0371) [inline reward/staking_info.go:66]:
```suggestion
func (st *StakingInfo) MarshalJSON() ([]byte, error) {
```

Comment (aidan-kwon) [reply reward/staking_info.go:66]: MarshalJSON need to be a struct, not a pointer. Please see https://groups.google.com/g/golang-nuts/c/gW53LC_ZC98

**Review by kjeom — APPROVED** [review]

**Comment by aidan-kwon** [comment]
@ethan-kr @ian0371 @blukat29 PTAL

**Review by blukat29 — APPROVED** [review]

File: reward/staking_info.go (hunk 4 above)
Comment (ian0371) [reply reward/staking_info.go:66]: I see. Then, there are lots of MarshalJSON with pointers in the klaytn repo that need to get fixed. For the reference, here's to see that it doesn't work:

```go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
)

type A struct{ a int }

func (a *A) MarshalJSON() ([]byte, error) {
	fmt.Println(a.a)
	return json.Marshal(a.a)
}

func main() {
	{
		x := interface{}(A{a: 10})
		fmt.Println(x.(A).a) // This works
		//fmt.Println(&x.(A).a) // This doesn't

		t := reflect.TypeOf(x)
		fmt.Println("Num methods", t.NumMethod())

		b, err := json.Marshal(x)
		fmt.Println(string(b), err)
	}
	fmt.Println("==================")
	{
		x := A{a: 10}
		t := reflect.TypeOf(x)
		fmt.Println("Num methods", t.NumMethod())

		b, err := json.Marshal(&x)
		fmt.Println(string(b), err)
	}
	fmt.Println("==================")
	{
		x := &A{a: 10}
		t := reflect.TypeOf(x)
		fmt.Println("Num methods", t.NumMethod())

		for i := 0; i < t.NumMethod(); i++ {
			fmt.Println("...", t.Method(i).Name)
		}

		b, err := json.Marshal(&x)
		fmt.Println(string(b), err)
	}
}
```

**Review by ian0371 — APPROVED** [review]

**Review by ethan-kr — APPROVED** [review]

**Pull request merged**